### ISSUE TITLE
Unauthorized REST requests are retried forever when OAuth is used

### ISSUE THREAD

**Issue by lenaschoenburg** [camunda/camunda#26012]
When making a REST request that results in a 401 Unauthorized, the Java client automatically retries the request forever, without any backoff when authentication via OAuth is used. 

This is because the `retryPredicate` in `ApiCallback` is `OAuthCredentialsProvider#shouldRetryRequest`. There are two angles to this:
1. The `ApiCallback` should probably have a retry counter, just like we introduced in https://github.com/camunda/camunda/pull/20088 for gRPC.
2. `OAuthCredentialsProvider#shouldRetryRequest` should not blindly request new access tokens on every call, this potentially overloads the OAuth server with too many requests.

:thought_balloon: This  is also related to https://github.com/camunda/camunda/issues/26010 where _missing permissions_
would result in a 401 which `OAuthCredentialsProvider#shouldRetryRequest` interprets as a retriable request.

**Comment by lenaschoenburg** [comment]
The fix that was implemented does not work correctly because it stores the retry counter in the `ApiCallback` object which gets recreated on every retry 😅 https://github.com/camunda/camunda/pull/26442

**Comment by megglos** [comment]
> The fix that was implemented does not work correctly because it stores the retry counter in the `ApiCallback` object which gets recreated on every retry 😅 [#26442](https://github.com/camunda/camunda/pull/26442)

Thanks for raising, looking at it seems we also preliminary closed this issue, as https://github.com/camunda/camunda/pull/26442 was only addressing 
> 1. The `ApiCallback` should probably have a retry counter, just like we introduced in https://github.com/camunda/camunda/pull/20088 for gRPC.

but not

> 2. `OAuthCredentialsProvider#shouldRetryRequest` should not blindly request new access tokens on every call, this potentially overloads the OAuth server with too many requests.

Following up on that. Do mean with "blindly" that it it generally retries on 401s, which effectively leads to using all tries in case the token obtained is generally not valid for the API. (e.g. wrong issuer or audience)

**Comment by megglos** [comment]
Effectively the cause for the `ApiCallback` recreation is the [retryAction calling sendRequest again](https://github.com/camunda/camunda/blob/main/clients/java/src/main/java/io/camunda/client/impl/http/HttpClient.java#L278) 😬

**Comment by lenaschoenburg** [comment]
> Do mean with "blindly" that it it generally retries on 401s, which effectively leads to using all tries in case the token obtained is generally not valid for the API. (e.g. wrong issuer or audience)

Exactly. If the token does not grant necessary permissions and leads to a 401, the credential provider always tries to retrieve a token. At least using Keycloak, this token is always a new one, which is interpreted as a signal to retry the request.

**Comment by ChrisKujawa** [comment]
Related with https://github.com/camunda/camunda/issues/34597

**Comment by megglos** [comment]
@nicpuppa please take this one, so we can fix this for good 💪
